cpp/src/arrow/ipc/adapter.cc

```diff
@@ -149,10 +149,19 @@ class RowBatchWriter {
   }
 
   // This must be called after invoking AssemblePayload
-  int64_t DataHeaderSize() {
-    // TODO(wesm): In case it is needed, compute the upper bound for the size
-    // of the buffer containing the flatbuffer data header.
-    return 0;
+  Status DataHeaderSize(int64_t* size) {
+    // emulates the behavior of Write without actually writing
+    int64_t offset = 0;
```

Comment:
Two possible suggestions for reducing code duplication:
1.  Just pass the MockMemory object to write in GetRowBatchSize() and use that to compute the total size (this makes the existing unit test a little less useful, you would actually want to verify the entire span of writes done by the Write method) and would change my comment below about where to place the mock memory object.
2.  change Write to WritePrivate, and have an additional out parameter that is either data header size (int) or the buffer containing the written data header (it would still require passing in the MockMemory object.